timemachine/fe/single_topology_v3.py

```diff
@@ -0,0 +1,539 @@
+    for a_idx, a_param in enumerate(mol_a_nbpl_params):
+        mol_c_nbpl_params[a_to_c[a_idx]] = a_param
+
+    bond_potential = potentials.HarmonicBond(mol_c_bond_idxs).bind(mol_c_bond_params)
+    angle_potential = potentials.HarmonicAngle(mol_c_angle_idxs).bind(mol_c_angle_params)
+    torsion_potential = potentials.PeriodicTorsion(mol_c_torsion_idxs).bind(mol_c_torsion_params)
+    nonbonded_potential = potentials.NonbondedPairList(
+        mol_c_nbpl_idxs,
+        mol_a_nbpl.get_rescale_mask(),
+        mol_a_nbpl.get_beta(),
+        mol_a_nbpl.get_cutoff(),
+    ).bind(mol_c_nbpl_params)
+
+    return system.VacuumSystem(bond_potential, angle_potential, torsion_potential, nonbonded_potential)
+
+
+def find_dummy_groups_and_anchors(mol_a, mol_b, core_a, core_b):
+    """
+    Find dummy groups for mol_b and appropriate bond/angle atoms in mol_a to
+    build restraints off of.
+
+    Parameters
+    ----------
+    mol_a: Chem.Mol
+        Fully interacting molecule
+
+    mol_b: Chem.Mol
+        Molecule that will provide dummy atoms
+
+    core_a: list of int
+        Core atoms in mol_a
+
+    core_b: list of int
+        Core atoms in mol_b
+
+    Returns
+    -------
+    set, list
+        Returns a set of dummy_groups, and a listo 2-tuples (j,k). j is the root_anchor atom,
+        and k is a neighboring core atom.
+
+    """
+    bond_idxs_b = utils.get_romol_bonds(mol_b)
+    dummy_groups_b = identify_dummy_groups(bond_idxs_b, core_b)
+
+    core_b_to_a = dict()
+    for ca, cb in zip(core_a, core_b):
+        core_b_to_a[cb] = ca
+
+    bond_idxs_a = utils.get_romol_bonds(mol_a)
+    jks = []
+
+    for dg in dummy_groups_b:
+        dummy_atom = list(dg)[0]
+        root_anchors = identify_root_anchors(bond_idxs_b, core_b, dummy_atom)
+
+        if len(root_anchors) == 0:
+            assert 0, "Disconnected dummy group"
+
+        if len(root_anchors) > 1:
+            # Consider the following situation:
+            # D0.D1
+            # .  .  where (.) is the dummy bond
+            # C0-C1
+            #
+            # One of (C0.D0), (D0.D1), (C1.D1) dummy bonds needs to be broken in order to maintain factorizability.
+            # This is a little arbitrary, but some choices are probably more efficient than others:
+            #
+            # hard     easy
+            # D0.D1    D0 D1
+            #    .     .  .
+            # C0-C1    C0-C1
+            #
+            # the lhs is more difficult because it has significantly more phase space that can be
+            # sampled than the fused case, but it's not super obvious how to best detect this.
+            # So instead, we will pick a random anchor atom. One possible solution later on is to
+            # minimize the number of rotatable bonds?
+            warnings.warn(
+                f"Multiple root anchors {root_anchors} found for dummy group: {dg}, picking a random anchor.",
+                MultipleAnchorWarning,
+            )
+
+        jk = None
```

Comment:
nit: jk -> angle_jk to make it more clear what this is for

Reply:
willfix